vulnerabilities/package_managers.py

```diff
@@ -489,3 +510,86 @@ async def fetch(self, pkg, session):
+
+    @staticmethod
+    def trim_url_path(url_path: str) -> Optional[str]:
+        """github advisories for golang is using package names(e.g. https://github.com/advisories/GHSA-jp4j-47f9-2vc3), yet goproxy works with module names(see https://golang.org/ref/mod#goproxy-protocol).
+        this method removes the last part of a package path, and returns the remaining as the module name.
+        """
+        # some advisories contains this prefix in package name, e.g. https://github.com/advisories/GHSA-7h6j-2268-fhcm
+        if url_path.startswith("https://pkg.go.dev/"):
+            url_path = url_path.removeprefix("https://pkg.go.dev/")
+        parts = url_path.split("/")
+        if len(parts) >= 2:
+            return "/".join(parts[:-1])
+        else:
+            return None
+
+    @staticmethod
+    def escape_path(path: str) -> str:
+        """escape uppercase in module/version name"""
+        escaped_path = ""
+        for c in path:
+            if c >= "A" and c <= "Z":
+                # replace uppercase with !lowercase
+                escaped_path += "!" + chr(ord(c) + ord("a") - ord("A"))
+            else:
+                escaped_path += c
+        return escaped_path
+
+    @staticmethod
+    async def parse_version_info(
+        version_info: str, escaped_pkg: str, session: ClientSession
+    ) -> Optional[Version]:
+        v = version_info.split()
+        if len(v) > 0:
+            value = v[0]
+            if len(v) > 1:
+                release_date = parse_datetime(v[1])
+            else:
+                escaped_ver = GoproxyVersionAPI.escape_path(value)
+                try:
+                    response = await session.request(
+                        method="GET",
+                        url=f"https://proxy.golang.org/{escaped_pkg}/@v/{escaped_ver}.info",
+                    )
+                    resp_json = await response.json()
+                    release_date = parse_datetime(resp_json.get("Time", ""))
+                except:
+                    release_date = None
+            return Version(value=value, release_date=release_date)
+        return None
+
+    async def fetch(self, pkg: str, session: ClientSession):
+        # escape uppercase in module path
+        escaped_pkg = GoproxyVersionAPI.escape_path(pkg)
+        trimmed_pkg = pkg
+        resp_text = None
+        while escaped_pkg is not None:
+            url = f"https://proxy.golang.org/{escaped_pkg}/@v/list"
+            try:
+                response = await session.request(method="GET", url=url)
+                resp_text = await response.text()
+            except HTTPGone:
+                escaped_pkg = GoproxyVersionAPI.trim_url_path(escaped_pkg)
+                trimmed_pkg = GoproxyVersionAPI.trim_url_path(trimmed_pkg) or ""
+                continue
+            break
+        if resp_text is None or escaped_pkg is None or trimmed_pkg is None:
+            print("error fetch versions from goproxy: ", pkg)
```

Comment:
nit: "error while fetching versions for {pkg} from goproxy" would be better err message ?